plugin-c/crypto.cpp

```diff
@@ -0,0 +1,253 @@
+  if (!secp256k1_ecdsa_recoverable_signature_parse_compact(ctx, &sig, sigArr, v_long - 27)) {
+    return hexEncode(nullptr, 0);
+  }
+  secp256k1_pubkey key;
+  if (!secp256k1_ecdsa_recover(ctx, &key, &sig, str->data)) {
+    return hexEncode(nullptr, 0);
+  }
+  unsigned char serialized[65];
+  size_t len = sizeof(serialized);
+  secp256k1_ec_pubkey_serialize(ctx, serialized, &len, &key, SECP256K1_EC_UNCOMPRESSED);
+  struct string *result = (struct string *)koreAlloc(sizeof(struct string) + 64);
+  result->b.len = 64;
+  memcpy(result->data, serialized+1, 64);
+  return result;
+}
+
+struct g1point {
+  struct blockheader h;
+  mpz_ptr x;
+  mpz_ptr y;
+};
+
+struct g2point {
+  struct blockheader h;
+  mpz_ptr x0;
+  mpz_ptr x1;
+  mpz_ptr y0;
+  mpz_ptr y1;
+};
+
+struct inj {
+  struct blockheader h;
+  void *data;
+};
+
+struct kseq {
+  struct blockheader h;
+  inj *inj;
+  void *dotk;
+};
+
+bool bn128_initialized = false;
+
+extern "C++" {
+
+void initBN128() {
+  if (bn128_initialized) {
+    return;
+  }
+  alt_bn128_pp::init_public_params();
+  inhibit_profiling_info = true;
+  bn128_initialized = true;
+}
+
+alt_bn128_G1 getPoint(g1point *pt) {
+  if (mpz_cmp_ui(pt->x, 0) == 0 && mpz_cmp_ui(pt->y, 0) == 0) {
+    return alt_bn128_G1::zero();
+  }
+  mpz_t mod;
+  mpz_init(mod);
+  alt_bn128_Fq::mod.to_mpz(mod);
+  if (mpz_cmp(pt->x, mod) >= 0 || mpz_cmp(pt->y, mod) >= 0) {
+    throw std::invalid_argument("not a member of the field");
+  }
+  mpz_clear(mod);
+  auto x = bigint<alt_bn128_q_limbs>(pt->x);
+  auto y = bigint<alt_bn128_q_limbs>(pt->y);
+  auto z = bigint<alt_bn128_q_limbs>(1);
+  return alt_bn128_G1{x, y, z};
+}
+
+alt_bn128_G2 getPoint(g2point *pt) {
+  if (mpz_cmp_ui(pt->x0, 0) == 0 && mpz_cmp_ui(pt->x1, 0) == 0 
+      && mpz_cmp_ui(pt->y0, 0) == 0 && mpz_cmp_ui(pt->y1, 0) == 0) {
+    return alt_bn128_G2::zero();
+  }
+  mpz_t mod;
+  mpz_init(mod);
+  alt_bn128_Fq::mod.to_mpz(mod);
+  if (mpz_cmp(pt->x0, mod) >= 0 || mpz_cmp(pt->x1, mod) >= 0
+      || mpz_cmp(pt->y0, mod) >= 0 || mpz_cmp(pt->y1, mod) >= 0) {
+    throw std::invalid_argument("not a member of the field");
+  }
+  mpz_clear(mod);
+  auto x = alt_bn128_Fq2{bigint<alt_bn128_q_limbs>(pt->x0), bigint<alt_bn128_q_limbs>(pt->x1)};
+  auto y = alt_bn128_Fq2{bigint<alt_bn128_q_limbs>(pt->y0), bigint<alt_bn128_q_limbs>(pt->y1)};
+  auto z = alt_bn128_Fq2::one();
+  return alt_bn128_G2{x, y, z};
+}
+
+g1point *projectPoint(uint64_t hdr, alt_bn128_G1 pt) {
+  mpz_ptr x, y;
+  if (pt.is_zero()) {
+    x = (mpz_ptr) malloc(sizeof(*x));
+    y = (mpz_ptr) malloc(sizeof(*y));
+    mpz_init_set_ui(x, 0);
+    mpz_init_set_ui(y, 0);
+  } else {
+    pt.to_affine_coordinates();
+    x = (mpz_ptr) malloc(sizeof(*x));
+    y = (mpz_ptr) malloc(sizeof(*y));
+    mpz_init(x);
+    mpz_init(y);
+    pt.X.as_bigint().to_mpz(x);
+    pt.Y.as_bigint().to_mpz(y);
+  }
+  struct g1point *g1pt = (struct g1point *)koreAlloc(sizeof(struct g1point));
+  g1pt->h.len = hdr;
```

Comment:
The naming looks a bit suspect here. The value is called `hdr`, and I think it's actually a general header tag, but the field is just named `len`. This should probably be addressed while moving some of the declarations into headers provided by the llvm backend. I seriously doubt we will continue indefinitely to use the full `uint64_t` for length instead of taking some bits for other purposed and using a shorter (bit)field.